### DISCUSSION TITLE
How to write proper v-if for conditional rendering

### DISCUSSION THREAD

**Discussion by alpeshvmodi** [nuxt/nuxt#16761] [Questions] [answered]
Hello everyone,

I am trying to display components on a conditional basis as follows.

```
    <div v-if="deviceType == 1">
      <Component1 />
    </div>
    <div v-if="deviceType == 2">
      <Component2 />
    </div>
<script setup>
  const deviceType = ref(null);
  const isMobile =
    /(Android|webOS|iPhone|iPad|iPod|tablet|BlackBerry|Mobile|Windows Phone)/i.test(
      headers['user-agent']
    );
  if (isMobile) {
    console.log('----setting 2');
    deviceType = ref(2);
  } else {
    console.log('----setting 1');
    deviceType = ref(1);
  }
  console.log('-----isMobile--', isMobile, deviceType.value);
</script>
```

When I run on a mobile browser, server-side I get values of `isMobile` and `deviceType.value` as desired. 
But when the same code evaluates in the browser, I am getting the wrong values of both as `headers` is not present there thus `headers['user-agent']` is `undefined`.

This results in the wrong component displayed with a Hydration error in the browser.

Any help is appreciated. 

Thanks Already,

**Comment by sevillaarvin** [comment]
How about using `useAsyncData`? https://github.com/nuxt/framework/discussions/4493#discussioncomment-2607854

**Reply by alpeshvmodi** [reply]
I am not sure, how to use useAsyncData to display one component conditionally during the first paint of content. Will still give it a try.

**Comment by alpeshvmodi** [comment]
Hey @sevillaarvin, Did you mean to say as below? It behaves the same. The `deviceType.value` is printed correctly in the server log, while in the browser console it appears wrong. The result is, the component appears for a moment and then disappears. I am stuck.

```
<div v-if="deviceType == 1">
    <Component1 />
</div>
<div v-if="deviceType == 2">
    <Component2 />
</div>
<script setup>
  const headers = useRequestHeaders();
  let deviceType = ref(null);
  await useAsyncData('key', async (app) => {
    const isMobile =
      /(Android|webOS|iPhone|iPad|iPod|tablet|BlackBerry|Mobile|Windows Phone)/i.test(
        headers['user-agent']
      );
    if (isMobile) {
      deviceType = ref(2);
    } else {
      deviceType = ref(1);
    }
  });
  console.log('-----isMobile-999-', deviceType.value);
</script>
```

**Comment by bot08** [comment, marked as answer]
You should process on the server side through headers, and on the client side through window.navigator.userAgent. Or transfer the deviceType to the client side without re-defining. (It must be saved via useState)

**Reply by alpeshvmodi** [reply]
Actually, I don't want to execute it on the client side. My actual requirement is to display either component1 or component2 based on the condition. But somehow above code executes both sides automatically.
Thanks for the tip @bot08, I will try out `useState`.

**Reply by alpeshvmodi** [reply]
Hey @bot08, the `useState` worked.
Big Thanks to you.!..

**Reply by sevillaarvin** [reply]
Nice, can you show how you did it?

**Reply by alpeshvmodi** [reply]
Sure @sevillaarvin,

I used it as follows, and my problem was solved.
```
<div v-if="deviceType == 1">
    <Component1 />
</div>
<div v-if="deviceType == 2">
    <Component2 />
</div>
<script setup>
  const headers = useRequestHeaders();
  const deviceType = useState('deviceType', () => {
    if (
      /(Android|webOS|iPhone|iPad|iPod|tablet|BlackBerry|Mobile|Windows Phone)/i.test(
        headers['user-agent']
      )
    ) {
      return ref(2);
    } else {
      return ref(1);
    }
  });
</script>
```
I hope that helps :)
